src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/AutoInsert/AutoInsertService.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Frozen;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Razor.AutoInsert;
+
+internal class AutoInsertService(IEnumerable<IOnAutoInsertProvider> onAutoInsertProviders) : IAutoInsertService
+{
+    private readonly ImmutableArray<IOnAutoInsertProvider> _onAutoInsertProviders = onAutoInsertProviders.ToImmutableArray();
+
+    public static FrozenSet<string> HtmlAllowedAutoInsertTriggerCharacters { get; }
+        = new string[] { "=" }.ToFrozenSet(StringComparer.Ordinal);
+    public static FrozenSet<string> CSharpAllowedAutoInsertTriggerCharacters { get; }
+        = new string[] { "'", "/", "\n" }.ToFrozenSet(StringComparer.Ordinal);
+
+    private readonly ImmutableArray<string> _triggerCharacters = CalculateTriggerCharacters(onAutoInsertProviders);
+
+    private static ImmutableArray<string> CalculateTriggerCharacters(IEnumerable<IOnAutoInsertProvider> onAutoInsertProviders)
+    {
+        using var builder = new PooledArrayBuilder<string>();
+        foreach (var provider in onAutoInsertProviders)
+        {
+            var triggerCharacter = provider.TriggerCharacter;
+            if (!builder.Any(character => character == triggerCharacter))
```

Comment:
I switched to using PooledArrayBuilder per Dustin's suggestion, which doesn't contain Contains (unless I'm missing something?)

I didn't act on "improving performance" here because it's a fixed-length 2-element list.

Since there is so much concern about it, I went ahead and switched to use HashSet to dedupe a 2-element list as suggested.

Reply:
Since you already mentioned you will do a clean-up pass of your own, you are welcome to modify this if still not to your liking. It's not going to hurt my feelings. :)

Reply:
> which doesn't contain Contains 

Thanks, that explains it